### PR TITLE
feat: add vesting smart contract example

### PR DESCRIPTION
**Pull request by 0xf333**
## Description
Vesting smart contract implementation.

## What This Enables
 - Linear Release Mechanism: Tokens or assets are distributed over a set duration, preventing sudden dumps.
 - Team /Investor: Incentive for long-term commitment with token allocations released over time.
 - Flexibility: Can be tailored to various vesting schedules.
 
## Impacts
An example of how to implement a vesting contract with Ink.

.

### CODE REVIEW

**Comment by 0xf333** [comment]
@athei

File: vesting/lib.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+                duration_time,
+                start_time,
+                beneficiary,
+                owner,
+                releasable_balance,
+                released_balance,
+            })
+        }
+
+        // This returns current block timestamp.
```
Comment (SkymanOne) [inline vesting/lib.rs:68]: Add those as docs with `///`

File: vesting/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "vesting"
+version = "0.1.0"
+authors = ["[] <[]>"]
```
Comment (SkymanOne) [inline vesting/Cargo.toml:4]: Please add authors like here https://github.com/paritytech/ink-examples/blob/3254a292b5cf13b087f92f2d02dc4bcbe847e71d/call-runtime/Cargo.toml#L4C1-L4C1
You can also leave yourself in the list

File: vesting/lib.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    /// The beneficiary cannot be the zero address.
+    ///
+    impl VestingContract {
+        #[ink(constructor, payable)]
+        pub fn new(beneficiary: AccountId, duration_time_in_sec: Timestamp) -> Result<Self, Error> {
+            if beneficiary == AccountId::from([0x0; 32]) {
+                return Err(Error::InvalidBeneficiary);
+            }
+            let duration_time = duration_time_in_sec * 1000;
+            let start_time = Self::env().block_timestamp();
```
Comment (SkymanOne) [inline vesting/lib.rs:53]: I feel like this is better be implemented in terms of block numbers. Given average block time of ~6s, let's say, we can calculate the duration of vesting period in the number of block.

**Review by SkymanOne — COMMENTED** [review]
Looks very good to me, although some changes are needed. 
I also suggest you introduce a trait and implement it. The reason is you might want to vest non-native tokens, like of psp22 standard, and/or through chain extensions: https://github.com/paritytech/ink-examples/tree/main/psp22-extension

File: vesting/lib.rs (hunk 1 above)
Comment (0xf333) [reply vesting/lib.rs]: done

File: vesting/Cargo.toml (hunk 2 above)
Comment (0xf333) [reply vesting/Cargo.toml]: done

File: vesting/lib.rs (hunk 3 above)
Comment (0xf333) [reply vesting/lib.rs:55]: Hi @SkymanOne  
Thanks for the feedback, I definitely see your point and it makes sense. 
However, I've once [came across this article](https://consensys.github.io/smart-contract-best-practices/development-recommendations/solidity-specific/timestamp-dependence/#avoid-using-blocknumber-as-a-timestamp) that recommends timestamps for time-dependent logic, as I did with `vesting_schedule`; Could you please take a look at the article and let me know if you still prefer that I use block numbers instead?

Let me know

**Comment by 0xf333** [comment]
> The reason is you might want to vest non-native tokens, like of psp22 standard, and/or through chain extensions: https://github.com/paritytech/ink-examples/tree/main/psp22-extension

Absolutely, I'm planing to do it once [the PSP22 example is completed](https://github.com/paritytech/ink-examples/tree/main/psp22-extension#disclaimer).

It will be a separated pull request to add the PSP22 support on this vesting contract.

File: vesting/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "vesting"
+version = "0.1.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2021"
+
+[dependencies]
+ink = { version = "4.2.0", default-features = false }
```
Comment (SkymanOne) [inline vesting/Cargo.toml:8]:
```suggestion
ink = { version = "4.2", default-features = false }
```

File: vesting/lib.rs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            } else {
+                return (total_allocation
+                    * (current_time - self.start_time()) as Balance)
+                    / self.duration_time() as Balance;
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
```
Comment (SkymanOne) [inline vesting/lib.rs:209]: Can you please also add a couple of E2E tests?

**Review by SkymanOne — CHANGES_REQUESTED** [review]

File: vesting/lib.rs (hunk 5 above)
Comment (0xf333) [reply vesting/lib.rs:209]: Thanks for the feedback, I will proceed with this request today evening

File: vesting/Cargo.toml (hunk 4 above)
Comment (0xf333) [reply vesting/Cargo.toml:8]: Thanks for the feedback, I will proceed with this request today evening

Comment (0xf333) [reply vesting/Cargo.toml]: done

File: vesting/lib.rs (hunk 5 above)
Comment (0xf333) [reply vesting/lib.rs:209]: > Can you please also add a couple of E2E tests?

I'm noticing that the substrate-contracts-node doesn't yet offer an interval mining feature similar to Hardhat for eth chain development; this makes E2E testing challenging for time-based contracts, as is the case here.
Once this feature is available, I'm open to revisiting this. I could then submit an additional pull request to enhance this contract with a few E2E tests.
For now, the current unit tests accurately reflecting the contract's behavior on testnets, such as Rococo and Aleph Zero, ect...

Please let me know what's your take on this.

**Review by SkymanOne — APPROVED** [review]

**Comment by 0xf333** [comment]
Thanks

**Pull request merged**